manager/dispatcher/orphans.go

```diff
@@ -0,0 +1,121 @@
+package dispatcher
+
+import (
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+	"golang.org/x/net/context"
+)
+
+var (
+	// downNodeCheckInterval specifies the ferquency at which the
+	// down nodes are scanned.
+	downNodeCheckInterval = 5 * time.Minute
+	// downNodeOrphanThreshold specifies the how long before the
+	// tasks that belong to down nodes are moved to ORPHANED
+	// state.
+	downNodeOrphanThreshold = 24 * time.Hour
+	// downNodeCheckMulti calculates the multiple so that the
+	// threshold logic doesn't depend on wall clock but instead
+	// calculates the thershold expiry as a multiple.
+	downNodeCheckMulti = (downNodeOrphanThreshold / downNodeCheckInterval)
+)
+
+// downNodeStore defines a type to store the nodes which are currently
+// down.
+type downNodeStore struct {
+	nodes map[string]int
+	store *store.MemoryStore
+	sync.Mutex
+}
+
+func newDownNodeStore(store *store.MemoryStore) *downNodeStore {
+	return &downNodeStore{
+		nodes: make(map[string]int),
+		store: store,
+	}
+}
+
+func (dn *downNodeStore) add(nodeID string) {
+	dn.Lock()
+	if _, ok := dn.nodes[nodeID]; ok {
+		dn.Unlock()
+		return
+	}
+
+	dn.nodes[nodeID] = 0
+	dn.Unlock()
+}
+
+func (dn *downNodeStore) remove(nodeID string) {
+	dn.Lock()
+	delete(dn.nodes, nodeID)
+	dn.Unlock()
+}
+
+func (dn *downNodeStore) run(ctx context.Context) {
+	t := time.NewTicker(downNodeCheckInterval)
+
+	for {
+		var candidateNodes []string
+		select {
+		case <-t.C:
+			dn.Lock()
+			for nodeID, count := range dn.nodes {
+				// This node has been down for too
+				// long. Add it to the candidateNodes
+				// list so that we can move its tasks
+				// to ORPHANED state and reap
+				// them. There is no reason to keep
+				// the node in downNodeStore either
+				// because after this, this node
+				// should not have any tasks which are
+				// not in ORPHANED state or reaped
+				// until the next time the node comes
+				// alive.
+				if count > int(downNodeCheckMulti) {
+					candidateNodes = append(candidateNodes, nodeID)
+					delete(dn.nodes, nodeID)
+					continue
+				}
+
+				dn.nodes[nodeID] = count + 1
+			}
+			dn.Unlock()
+		case <-ctx.Done():
+			return
+		}
+
+		for _, nodeID := range candidateNodes {
+			_, err := dn.store.Batch(func(batch *store.Batch) error {
+				return batch.Update(func(tx store.Tx) error {
```

Comment:
Make one call to `batch.Update` per task to update. If there are lots of tasks on one node, the size of the `Update` could exceed the raft transaction size limit.